### PR TITLE
refactor: remove drop_actors rpc and simplify barrier worker state

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Extension of https://github.com/risingwavelabs/risingwave/pull/18270.

In https://github.com/risingwavelabs/risingwave/pull/18270, we spawn actor in inject barrier, and there is no not-started actor any more. In this PR, we further simplify the actor state of local barrier worker. The `NotStarted` actor status is removed, and all actor related info, including join handles, are all managed in the `InflightActorState` and `StreamActorManagerState` is removed.

The `drop_actors` rpc is removed, because there is no need to use it. The drop actor info is carried in the barrier, and the info of the dropped actor can be removed when the barrier is completed locally.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 1]
```diff
@@ -416,32 +424,27 @@ impl PartialGraphManagedBarrierState {
     }
 }
 
-pub(super) struct ManagedBarrierState {
+pub(crate) struct ManagedBarrierState {
     pub(super) actor_states: HashMap<ActorId, InflightActorState>,
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:428]: May I know what "inflight" in `InflightActor` is for?

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 2]
```diff
@@ -550,53 +561,104 @@ impl ManagedBarrierState {
...
+                        partial_graph_id,
+                        barrier,
+                        join_handle,
+                        monitor_join_handle
+                    )
+                )
+                .is_ok());
+        }
+
+        if cfg!(test) {
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:621]: Can you add some comments?

File: src/stream/src/task/barrier_manager/managed_state.rs [hunk 3]
```diff
@@ -610,9 +672,12 @@ impl ManagedBarrierState {
     ) -> impl Future<Output = (PartialGraphId, u64)> + '_ {
         poll_fn(|cx| {
             for (partial_graph_id, graph_state) in &mut self.graph_states {
-                if let Poll::Ready(epoch) = graph_state.poll_next_completed_epoch(cx) {
+                if let Poll::Ready(barrier) = graph_state.poll_next_completed_barrier(cx) {
+                    if let Some(actors_to_stop) = barrier.all_stop_actors() {
+                        self.current_shared_context.drop_actors(actors_to_stop);
+                    }
```
Comment (BugenZhao) [inline src/stream/src/task/barrier_manager/managed_state.rs:678]: Can we also move this clean-up step into `collect -> is_finished`?

File: src/stream/src/task/barrier_manager/managed_state.rs (hunk 1 above)
Comment (wenym1) [reply src/stream/src/task/barrier_manager/managed_state.rs:428]: It has no special meaning. I guess the name was following the `InflightActorInfo` previously in the meta node global barrier manager.

File: src/stream/src/task/barrier_manager/managed_state.rs (hunk 3 above)
Comment (wenym1) [reply src/stream/src/task/barrier_manager/managed_state.rs:679]: At the beginning I implement it in this way, but later I realize that we also need to clean up the actor infos in other compute node, but the `collect -> is_finished` can only clean up the actor on this compute node. So now I changed to clean up here.

**Review by yezizp2012 — APPROVED** [review]
LGTM!

**Pull request merged**